pkg/gardenlet/controller/federatedseed/shootstate/common_sync.go

```diff
@@ -93,61 +88,65 @@ func (s *SyncController) createShootStateSyncReconcileFunc(ctx context.Context,
 		shootState := &gardencorev1alpha1.ShootState{ObjectMeta: metav1.ObjectMeta{Name: shootName, Namespace: projectNamespace}}
 
 		if _, err := controllerutil.CreateOrUpdate(ctx, s.k8sGardenClient.Client(), shootState, func() error {
-			shootState.OwnerReferences = []metav1.OwnerReference{*ownerReference}
 			return updateShootStateExtensionState(state, shootState, kind, name, purpose)
 		}); err != nil {
-			message := fmt.Sprintf("Shoot's %s %s extension state with name %s and purpose %v was NOT successfully synced: %v", shootName, kind, name, purpose, err)
+			message := fmt.Sprintf("Shoot's %s %s extension state with name %s and purpose %s was NOT successfully synced: %v", shootName, kind, name, purposeToString(purpose), err)
 			s.log.Error(message)
 			s.recorder.Event(shootState, corev1.EventTypeNormal, "ScheduledNextSync", message)
 			return reconcile.Result{}, err
 		}
 
-		message := fmt.Sprintf("Shoot's %s %s extension state with name %s and purpose %v was successfully synced", shootName, kind, name, purpose)
+		message := fmt.Sprintf("Shoot's %s %s extension state with name %s and purpose %s was successfully synced", shootName, kind, name, purposeToString(purpose))
 		s.log.Info(message)
 		s.recorder.Event(shootState, corev1.EventTypeNormal, "ScheduledNextSync", message)
 		return reconcile.Result{}, nil
 	}
 }
 
 func updateShootStateExtensionState(extensionState *runtime.RawExtension, shootState *gardencorev1alpha1.ShootState, kind string, name string, purpose *string) error {
-	i, currentExtensionState := gardencorev1alpha1helper.GetExtensionResourceState(shootState.Spec.Extensions, kind, &name, purpose)
-	if extensionState == nil && i > -1 {
-		lastElementIndex := len(shootState.Spec.Extensions) - 1
-		shootState.Spec.Extensions[i] = shootState.Spec.Extensions[lastElementIndex]
-		shootState.Spec.Extensions = shootState.Spec.Extensions[0:lastElementIndex]
+	list := gardencorev1alpha1helper.GetExtensionResourceStateList(shootState)
+	if extensionState == nil {
+		list.Delete(kind, &name, purpose)
+		gardencorev1alpha1helper.SetExtensionResourceStateList(shootState, list)
 		return nil
 	}
+	list.Upsert(&gardencorev1alpha1.ExtensionResourceState{
+		Kind:    kind,
+		Name:    &name,
+		Purpose: purpose,
+		State:   *extensionState,
+	})
+	gardencorev1alpha1helper.SetExtensionResourceStateList(shootState, list)
+	return nil
+}
+
+func purposeToString(purpose *string) string {
+	if purpose == nil {
+		return "<nil>"
+	}
+	return string(*purpose)
+}
 
-	if currentExtensionState == nil {
-		shootState.Spec.Extensions = append(shootState.Spec.Extensions, gardencorev1alpha1.ExtensionResourceState{
-			Kind:    kind,
-			Name:    &name,
-			Purpose: purpose,
-			State:   gardencorev1alpha1.ProviderConfig{RawExtension: *extensionState},
-		})
+func fromRequest(req reconcile.Request) (projectNamespace, shootName string) {
+	// If the resource is cluster scoped then we need to get the projectName and shootName from the resource's name
```

Comment:
Hmm, can you elaborate a bit here? The only resource where this is possible is the `Cluster` resource (which itself doesn't store any state that should be saved). Why do we need this?

Reply:
The backupentry is also cluster scoped.

Reply:
Okay, thank you. Maybe add this to the comment so that people can get an idea why this is needed.